zerver/lib/presence.py

```diff
@@ -42,95 +30,72 @@ def get_status_dicts_for_rows(all_rows: List[Dict[str, Any]],
-    for row in reversed(presence_rows):
-        if row['status'] == UserPresence.ACTIVE:
-            active_timestamp = datetime_to_timestamp(
-                row['timestamp'])
-            break
-
-    idle_timestamp = None
-    for row in reversed(presence_rows):
-        if row['status'] == UserPresence.IDLE:
-            idle_timestamp = datetime_to_timestamp(
-                row['timestamp'])
-            break
-
-    # Be stingy about bandwidth, and don't even include
-    # keys for entities that have None values.  JS
-    # code should just do a falsy check here.
+def get_modern_user_info(last_active_time: datetime.datetime,
+                         last_connected_time: datetime.datetime) -> Dict[str, Any]:
+    # TODO: Do further bandwidth optimizations to this structure.
     result = dict()
-
-    if active_timestamp is not None:
-        result['active_timestamp'] = active_timestamp
-
-    if idle_timestamp is not None:
-        result['idle_timestamp'] = idle_timestamp
-
+    result['active_timestamp'] = last_active_time
+    result['idle_timestamp'] = last_connected_time
     return result
 
-def get_legacy_user_info(presence_rows: List[Dict[str, Any]],
-                         mobile_user_ids: Set[int]) -> Dict[str, Any]:
-
-    # The format of data here is for legacy users of our API,
-    # including old versions of the mobile app.
-    info_rows = []
-    for row in presence_rows:
-        client_name = row['client__name']
-        status = UserPresence.status_to_string(row['status'])
-        dt = row['timestamp']
-        timestamp = datetime_to_timestamp(dt)
-        push_enabled = row['user_profile__enable_offline_push_notifications']
-        has_push_devices = row['user_profile__id'] in mobile_user_ids
-        pushable = (push_enabled and has_push_devices)
-
-        info = dict(
-            client=client_name,
-            status=status,
-            timestamp=timestamp,
-            pushable=pushable,
-        )
-
-        info_rows.append(info)
+def get_legacy_user_info(last_active_time: datetime.datetime,
+                         last_connected_time: datetime.datetime) -> Dict[str, Any]:
+    # Reformats the modern UserPresence data structure so that legacy
+    # API clients can still access presence data.
+    #
+    # We expect this code to remain mostly unchanged until we can delete it.
 
-    most_recent_info = info_rows[-1]
-
-    result = dict()
-
-    # The word "aggegrated" here is possibly misleading.
-    # It's really just the most recent client's info.
-    result['aggregated'] = dict(
-        client=most_recent_info['client'],
-        status=most_recent_info['status'],
-        timestamp=most_recent_info['timestamp'],
+    if timezone_now() - last_active_time > datetime.timedelta(minutes=2):
+        dt = last_active_time
+        status = UserPresence.LEGACY_STATUS_ACTIVE
+    else:
+        dt = last_connected_time
+        status = UserPresence.LEGACY_STATUS_IDLE
```

Comment:
This conditional is backwards, isn't it?  In any case, should this `minutes=2` be OFFLINE_THRESHHOLD_SECS?

Reply:
yeah, this was backwards, fixed now

Reply:
Unsure what timedelta is right here, I feel like there's a lot of different timedeltas and it's fairly confusing. Should we try to reduce the amount of different timedeltas we're using and unify this somehow?

I don't think OFFLINE_THRESHOLD_SECS is right here, because that was for categorizing a user as "offline", not "idle"?

Reply:
I think we don't even know why `OFFLINE_THRESHOLD_SECS` is a completely different value in `presence.js`, which isn't helping the situation either :neutral_face: 

(there's a TODO saying that):
```
# How long to wait before presence should treat a user as offline.
# TODO: Figure out why this is different from the corresponding
# value in static/js/presence.js.  Also, probably move it out of
# default_settings, since it likely isn't usefully user-configurable.
OFFLINE_THRESHOLD_SECS = 5 * 60
```

Reply:
So, the reason we have it in `default_settings` is that ideally we would have the option in the event of a users/load surge of making presence updates much slower (e.g. every 5 minutes rather than 1).  We used that a lot when first building the feature; I'm unsure if we're use it again, but that's probably why it's setup that way.